### PR TITLE
parallel assignment of evm circuit

### PR DESCRIPTION
**Pull request by lispc**
in this PR, assignment of evm circuit is split into 3 groups of regions:

1. real steps
2. padding EndBlock 
3. real final EndBlock and remaining assignments

only the (1)(2) need to be parallelized.

btw, the halo2 parallel assignment API is already very stable, so this time i am not going to use a env var to control whether to enable parallel assignment

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution.rs [hunk 1]
```diff
@@ -246,7 +247,7 @@ pub(crate) struct ExecutionConfig<F> {
     // EVM Circuit selector, which enables all usable rows.  The rows where this selector is
     // disabled won't verify any constraint (they can be unused rows or rows with blinding
     // factors).
-    q_usable: Selector,
+    q_usable: Column<Fixed>,
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution.rs:250]: if we still use selectors, halo2 MockProver will report err for cross region gates, which are needed for parallel assignment

File: zkevm-circuits/src/evm_circuit/execution/end_block.rs [hunk 2]
```diff
@@ -32,7 +34,24 @@ pub(crate) struct EndBlockGadget<F> {
     max_txs: Cell<F>,
     phase2_withdraw_root: Cell<F>,
     phase2_withdraw_root_prev: Cell<F>,
-    pub withdraw_root_assigned: std::cell::RefCell<Option<AssignedCell>>,
+    pub withdraw_root_assigned: Mutex<Option<AssignedCell>>,
```
Comment (kunxian-xia) [inline zkevm-circuits/src/evm_circuit/execution/end_block.rs:37]: why do you use `Mutex` here?

**Review by kunxian-xia — COMMENTED** [review]
LGTM except one question about the type change of `withdraw_root_assigned`.

File: zkevm-circuits/src/evm_circuit/execution.rs [hunk 3]
```diff
@@ -1217,15 +1355,10 @@ impl<F: Field> ExecutionConfig<F> {
         let final_withdraw_root_cell = self
             .end_block_gadget
             .withdraw_root_assigned
-            .borrow()
+            .lock()
```
Comment (noel2004) [inline zkevm-circuits/src/evm_circuit/execution.rs:1358]: It seems `withdraw_root_assigned` is assigned without cocurrent (inside of a `assign_region`, not `assign_regions`) so do we really need to protect it with mutex?

**Review by noel2004 — APPROVED** [review]
LGTM except some trivial questions

**Comment by lispc** [comment]
the reason why use Mutex, is due to assign_regions API need `Config` be of `Send` and `Sync`. So i have to wrap the withdraw root inside Mutex

**Pull request merged**